### ISSUE TITLE
System.Text.Json ignores [JsonConstructor] when class implements a generic read-only collection

### ISSUE THREAD

**Issue by ugumba** [dotnet/runtime#49699]
### Description

I have multiple (immutable) classes with getter-only properties and a matching `[JsonConstructor]`.
They also implement a generic read-only collection interface (e.g. `IReadOnlyList<T>` or `IReadOnlyDictionary<K,V>`), for ease of consumption elsewhere.
The collection interfaces are implemented in terms of the properties on the class.

An dumbed-down example:

```
        public class SillyList : IReadOnlyCollection<string>
        {
            [JsonConstructor]
            public SillyList(string item) { Item = item; }
            public string Item { get; }

            public int Count => 1;
            public IEnumerator<string> GetEnumerator() => new List<string> { Item }.GetEnumerator();
            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }

        [Fact]
        public void SerializeSillyList()
        {
            var data = new SillyList("yum");
            var json = JsonSerializer.Serialize(data);
            Assert.Equal("{\"Item\":\"yum\"}", json);  // fails: json == "{\"yum\"}"

            var back = JsonSerializer.Deserialize<SillyList>(json);  // throws NotSupportedException
            Assert.Equal("yum", back.Item);
        }
```
On serialization, System.Text.Json uses the collection interface - the properties are ignored.
On deserialization, it fails with NotSupportedException ("`The collection type 'SillyList' is abstract, an interface, or is read only`").

I would have assumed that the presence of `[JsonConstructor]` would be enough of a hint that System.Text.Json should ignore any implemented interfaces!  Newtonsoft.Json worked nicely.

I can find no way of overriding this behaviour, short of applying explicit `[JsonConverter]` on each and every one of these classes.  This would be a huge pain, as each converter needs to do a lot of manual work to override the problematic behaviour.

### Configuration

* .NET 5.0

### Regression?

Not per-se, but Newtonsoft.Json handled these classes quite gracefully.

**Comment by pinkfloydx33** [comment]
I hit a similar issue some time back, just because my class implemented IEnumerable<> which then broke the deserializer. Likely the same issue as discussed before.

https://github.com/dotnet/runtime/issues/46920

**Comment by ugumba** [comment]
Closed as duplicate of #46920.
